src/ephy-sync-service.c

```diff
@@ -0,0 +1,1417 @@
+{
+  StorageServerRequestAsyncData *data;
+
+  data = g_slice_new (StorageServerRequestAsyncData);
+  data->service = g_object_ref (service);
+  data->endpoint = g_strdup (endpoint);
+  data->method = method;
+  data->request_body = g_strdup (request_body);
+  data->modified_since = modified_since;
+  data->unmodified_since = unmodified_since;
+  data->callback = callback;
+  data->user_data = user_data;
+
+  return data;
+}
+
+static void
+storage_server_request_async_data_free (StorageServerRequestAsyncData *data)
+{
+  if (G_UNLIKELY (!data))
+    return;
+
+  g_object_unref (data->service);
+  g_free (data->endpoint);
+  g_free (data->request_body);
+  g_slice_free (StorageServerRequestAsyncData, data);
+}
+
+static void
+destroy_session_response_cb (SoupSession *session,
+                             SoupMessage *msg,
+                             gpointer     user_data)
+{
+  JsonParser *parser;
+  JsonObject *json;
+
+  if (msg->status_code == 200) {
+    LOG ("Session destroyed");
+    return;
+  }
+
+  parser = json_parser_new ();
+  json_parser_load_from_data (parser, msg->response_body->data, -1, NULL);
+  json = json_node_get_object (json_parser_get_root (parser));
+
+  g_warning ("Failed to destroy session: errno: %ld, errmsg: %s",
+             json_object_get_int_member (json, "errno"),
+             json_object_get_string_member (json, "message"));
+
+  g_object_unref (parser);
+}
+
+static gboolean
+ephy_sync_service_storage_credentials_is_expired (EphySyncService *self)
+{
+  g_return_val_if_fail (EPHY_IS_SYNC_SERVICE (self), TRUE);
+
+  if (self->storage_credentials_id == NULL || self->storage_credentials_key == NULL)
+    return TRUE;
+
+  if (self->storage_credentials_expiry_time == 0)
+    return TRUE;
+
+  /* Consider a 60 seconds safety interval. */
+  return self->storage_credentials_expiry_time < ephy_sync_utils_current_time_seconds () - 60;
+}
+
+static void
+ephy_sync_service_fxa_hawk_post_async (EphySyncService     *self,
+                                       const char          *endpoint,
+                                       const char          *id,
+                                       guint8              *key,
+                                       gsize                key_length,
+                                       char                *request_body,
+                                       SoupSessionCallback  callback,
+                                       gpointer             user_data)
+{
+  EphySyncCryptoHawkOptions *hoptions;
+  EphySyncCryptoHawkHeader *hheader;
+  SoupMessage *msg;
+  char *url;
+  const char *content_type = "application/json";
+
+  g_return_if_fail (EPHY_IS_SYNC_SERVICE (self));
+  g_return_if_fail (endpoint != NULL);
+  g_return_if_fail (id != NULL);
+  g_return_if_fail (key != NULL);
+  g_return_if_fail (request_body != NULL);
+
+  url = g_strdup_printf ("%s%s", MOZILLA_FXA_SERVER_URL, endpoint);
+  msg = soup_message_new (SOUP_METHOD_POST, url);
```

Comment:
`soup_session_queue_message` steals the reference on the SoupMessage and it unrefs the message for us after returning from the callback.